### ISSUE TITLE
Stream does not play correctly

### ISSUE THREAD

**Issue by malamalca** [mylisabox/flutter_mjpeg#14]
I am streaming raspicam via uv4l. The stream plays fine via Flutter Webview Component or Flutter VLC Player.

Trying to open strem with MJPEG widget produces half-rendered images, jittering, split images as seen in attached video.

Is it possible that --Boundary-- header preceding image produces streaming errors?

A link to a sample video and first part of my binary stream:
[video.mp4](https://user-images.githubusercontent.com/165478/121405629-9098b300-c95d-11eb-99e5-b645dc87bbe7.mp4)

**Comment by YowFung** [comment]
Hi @jaumard, it may be about `gaplessPlayback` property of `Image` Widget in your `Mjpeg` Widget. Set to `true` can solve it.

> https://github.com/mylisabox/flutter_mjpeg/blob/01c3cd4c83544ee07feb4b37a188d00740e4267a/lib/src/mjpeg.dart#L98

**Comment by malamalca** [comment]
I think the problem is the stream. It might be currupting data.

I captured binary stream inside updateStream() function and saved stream to file. First two or three images were ok, the rest of them were corrupted (partly rendered, vertically splitted,..).

After that I've captured stream with php (yes, I know..), saved stream to a file and extracted jpeg images. They were all ok.

**Comment by jaumard** [comment]
@malamalca when you captured the steam, did you capture the raw data coming out of the HTTP request or the data I parse ? 

Because if it's the raw data coming out of the HTTP I'm really not sure what I can do about it... it would mean the the default dart HTTP package have issues. Or am I missing something ?

**Comment by malamalca** [comment]
Ok, I've solved my issue by converting playback from async to sync-ed code.
I guess there is a race condition in stream.listen which is adding data to _carry buffer.

gaplessPlayback prevents a bit of flickering too.

```dart
void updateStream(BuildContext context, ValueNotifier<MemoryImage?> image, ValueNotifier<dynamic> errorState) {
    try {
      final request = Request("GET", Uri.parse(stream));
      _httpClient.send(request).then((response) {
        if (response.statusCode >= 200 && response.statusCode < 300) {
          var _carry = <int>[];
          _subscription = response.stream.listen((chunk) async {
            if (_carry.isNotEmpty && _carry.last == _trigger) {
              if (chunk.first == _eoi) {
                _carry.add(chunk.first);
                _sendImage(context, image, errorState, _carry);
                _carry = [];
                if (!isLive) {
                  dispose();
                }
              }
            }

            for (var i = 0; i < chunk.length - 1; i++) {
              final d = chunk[i];
              final d1 = chunk[i + 1];

              if (d == _trigger && d1 == _soi) {
                _carry.add(d);
              } else if (d == _trigger && d1 == _eoi && _carry.isNotEmpty) {
                _carry.add(d);
                _carry.add(d1);

                _sendImage(context, image, errorState, _carry);
                _carry = [];
                if (!isLive) {
                  dispose();
                }
              } else if (_carry.isNotEmpty) {
                _carry.add(d);
                if (i == chunk.length - 2) {
                  _carry.add(d1);
                }
              }
            }
          }, onError: (err) {
            try {
              errorState.value = err;
              image.value = null;
            } catch (ex) {}
            dispose();
          }, cancelOnError: true);
        } else {
          errorState.value =
              HttpException('Stream returned ${response.statusCode} status');
          image.value = null;
          dispose();
        }
      });
    } catch (error) {
      errorState.value = error;
      image.value = null;
    }
  }
```

**Comment by jaumard** [comment]
@malamalca please create a PR :) it will be easier for me know compare what you changed! But if it makes sense and fixes a bug, for sure I'll gladly merge your changes

**Comment by YowFung** [comment]
I also found relevant phenomena a few days ago: The MJPEG images can be displayed in macOS and iOS, but not in Android and Windows. I capture the binary data and save it as a JPG file, which has the same effect (macOS/iOS can, but Android/Windows can't).

Then I found that the binary data contained the HTTP protocol header. Like MJPEG frames, they all start with `FF D8 ...`.

![image](https://user-images.githubusercontent.com/26346560/121556465-7b4b8380-ca46-11eb-93d6-d0e722807210.png)

**Comment by YowFung** [comment]
In addition, there is one thing I don't quite understand.
> https://github.com/mylisabox/flutter_mjpeg/blob/01c3cd4c83544ee07feb4b37a188d00740e4267a/lib/src/mjpeg.dart#L136-L143

Do invalid image frames appear to be used?
I personally understand that it is correct to write like the following (I am not quite sure):
```dart
try {
    var isInvalid = false;
    await precacheImage(imageMemory, context, onError: (err, trace) {
        print("Invalid image: $err");
        isInvalid = true;
    });
    if (!isInvalid) {
        errorState.value = null;
        image.value = imageMemory;
    }
} catch (ex) {}
```

**Comment by jaumard** [comment]
Pushed the fix as 2.0.1
